src/browser-polyfill.js

```diff
@@ -361,18 +361,40 @@ if (typeof browser === "undefined") {
        *        yield a response. False otherwise.
        */
       return function onMessage(message, sender, sendResponse) {
-        let result = listener(message, sender);
+        let didCallSendResponse = false;
+
+        let wrappedSendResponse;
+        let sendResponsePromise = new Promise(resolve => {
+          wrappedSendResponse = function(response) {
+            didCallSendResponse = true;
+            resolve(response);
+          };
+        });
+
+        let result = listener(message, sender, wrappedSendResponse);
+
+        const isResultThenable = result !== true && isThenable(result);
+
+        // If the listener didn't returned true or a Promise, or called
+        // wrappedSendResponse synchronously, we can exit earlier
+        // because there will be no response sent from this listener.
+        if (result !== true && !isResultThenable && !didCallSendResponse) {
+          return false;
+        }
 
-        if (isThenable(result)) {
+        // If the listener returned a Promise, send the resolved value as a
+        // result, otherwise wait the promise related to the wrappedSendResponse
+        // callback to resolve and send it as a response.
+        if (isResultThenable) {
           result.then(sendResponse, error => {
             console.error(error);
-            sendResponse(error);
```

Comment:
Why did you remove error handling?

Reply:
Well, I did it because it doesn't seem that it actually reproduce the expected behavior (which should be to make the promise returned by the runtime.sendMessage, or tabs.sendMessage, on the other side to be rejected), but it adds an additional unexpected behavior: the rejected value is sent to the other side as the resolved value of the promise returned by sendMessage, which will be an empty object if the error was an instance of Error, or the rejected value if it is a plain object that can be successfully send over the messaging abstraction).

I guess that it could be safer to leave it there for now (given that it is how the polyfill currently behave) and evaluate if we can make it more similar to what we need in a follow up.

@Rob--W How that sounds to you?

As an additional side note: the only way I can think of right now to make it to behave more similar to how it does on Firefox is to add an additional custom static wrapper for the runtime.sendMessage/tabs.sendMessage side, so that we can send a "json representation" of the resolve and reject scenarios (and so basically turning the Error object into a plain json object and convert it back to an error on the other side) and reject the promise returned by runtime.sendMessage/tabs.sendMessage accordingly, e.g. something like this on the onMessage wrapper side:

```js
          result.then(msg => {
            // send the message value.
            sendResponse({msg});
          }, error => {
            console.error(error);
            // Send a JSON representation of the error if the rejected value
            // is an instance of error, or the object itself otherwise.
            sendResponse((error instanceof Error) ? {
              error: {
                isInstanceOfError: true,
                message: String(error),
                fileName: error.fileName,
                lineNumber: error.lineNumber,
              },
            } : {error});
```

and then on the wrapped `sendMessage` side something like:

```js
return new Promise((resolve, reject) => {
   const wrappedCallback = {msg, error}) => {
     if (error) {
       // Convert the JSON representation of the error back to an Error instance,
       // otherwise just reject with the error value as it is.
       if (error.isInstanceOfError) {
         reject(new Error(error.message, error.fileName, error.lineNumber));
       } else {
         reject(error);
       }
     } else {
       resolve(msg);
     }
  });

  ....
});
```

This has clearly its own drawbacks, e.g. if on the other side the extension is mixing the browser and chrome API namespaces (e.g. using the native chrome.runtime.sendMessage instead of the wrapped one) or if the message is being send to another extension (that may be not be using the polyfill).

Reply:
The error should be a rare case. So you could also just pass the message as-is, and wrap the error in an object with a very unusual key that is unlikely to conflict with a real message. If `isInstanceOfError` fits that bill, then use that. I would be more inclined to use a more verbose string from which it is obvious that the polyfill is the source.

Reply:
Perhaps something like `__mozWebExtPolyfill_isSerializedErrorObject` could work.